spec_01.md

```diff
@@ -0,0 +1,48 @@
+- Spec: 1
+- Title: Methods on libraries
+- Version: 0.1
+- Last-Modified: 2019-07-19
+- Authors: [James Bednar](),
+           [Julia Signell](jsignell@gmail.com),
+           [Jake Vanderplas]()
+- Status: Active
+- Type: Standards
+- Content-Type: text/markdown
+- Created: 2019-07-19
+
+## Abstract
+This spec describes a proposed minimal shared Python API for plotting libraries.
+
+This spec uses LIBRARY to refer to any specific visualization library.
+
+## If a method doesn't make sense
+If a given operation doesn't make sense for that library, then it can satisfy the spec by simply having that method return a message to that effect ("Unsupported: LIBRARY does not provide JSON output").
+
+## `.__spec_version__()` method
+Calling `.__spec_version__()` on the library should return a list of the specs that the library complies with. It should include the version of the spec that is being used.
+
+```python
+>>> LIBRARY.__spec_version__()
+{1: 0.1, 2: 0.1}
+```
+
+## Enable output from library
+Each library should provide a uniform way to enable itself in jupyter notebook & jupyterlab.
+
+> #### What currently exists
+> - matplotlib has `%matplotlib inline`
+> - bokeh has `bokeh.output_notebook()`
+> - altair has `alt.renderers.enable('notebook')`
+> - holoviews has `hv.extension('bokeh')`
+> - hvplot has `import hvplot.pandas`
+
+### Proposal 1 : magics
+Every library should define something like `%enable_LIBRARY`
```

Comment:
I think the spec should only list a single proposal.  Other proposals can be in issues and PRs on this repo, which if accepted would replace the previous proposal. In this case, I propose that we delete proposal 1 as not being sufficiently universal, because it unnecessarily ties the user code to be valid only in a Jupyter context.

Reply:
Right. The idea is that we choose Proposal 1 **OR** Proposal 2 before we merge this PR.

Reply:
Quick, choose proposal 2 before anyone objects! :-)

Reply:
The more I think about this I'm wondering if there are two concepts here or maybe one that depends on another. Could this be seen (at least for some libraries) as choosing a "backend"? Or choosing a "renderer"? For example, holoviews sets the extension to bokeh, is that the only way it can output to a notebook? Does bokeh mean it has to be in a notebook or could it be used to define the holoviews "backend" and then could save to disk instead of displaying in the notebook.

Vispy supports different GUI frameworks (PyQt5/Pyside2, Wx, glfw, etc) and one of those options is a notebook widget. One thing we've been talking about recently is how to **reliably* determine which backend is being used especially related to notebooks. For example, are we in a notebook but the user said they want to display in a PyQt5 window? We've ended up looking at some code that matplotlib has for determining what graphics backend should be used.

So I guess my question is what does enabling a library for output in a notebook mean? Should this spec define how backends are enabled/disabled/configured or is this specifically an interface for turning on notebook output no matter what that means internally to the library? I suppose the `output` keyword argument helps with that.

Reply:
No problem. For HoloViews, yes, choosing the backend is indeed conflated with enabling notebook output, by design -- the same call works in and out of a notebook, enabling notebook output (and choosing a backend) in the notebook, and simply choosing the backend outside of the notebook.  That way people can use the same .ipynb file as source code for a notebook and for a deployed dashboard in a non-Jupyter-based server.

But yes, this spec doesn't cover the case where you are in a notebook and want to generate a figure separately from a notebook.  Maybe it's ok that such a situation isn't covered, though?

Reply:
Yes, I think that's ok. Just wanted to make sure I was understanding this and that the scope for this interface is defined. It sounds like it should be specifically "enable output to a jupyter notebook no matter what that means" with logical defaults and keyword argument overrides that may depend on the library.